### ISSUE TITLE
fix: resolve typing inconsistencies and lint issues flagged by Copilot in PR #49

### ISSUE THREAD

**Issue by ktinubu** [tinaudio/synth-setter#51]
## Summary

Copilot's review of PR #49 flagged several categories of issues. This issue compiles the **unique, actionable** findings — deduplicated from ~37 raw comments — with analysis and suggested fixes.

Cross-ref: https://github.com/ktinubu/synth-permutations/pull/49

---

## 1. Batch type annotation mismatch (runtime-correct, type-wrong)

**Files affected:**
- `src/models/surge_ff_module.py` — `training_step` (L38), `validation_step` (L47), `test_step` (L58), `predict_step` (L69)
- `src/models/surge_flowvae_module.py` — `training_step` (L55), `validation_step` (L75), `predict_step` (L96)
- `src/models/surge_flow_matching_module.py` — `_train_step` (L117)

**Problem:** Lightning step methods annotate `batch` as `tuple[torch.Tensor, torch.Tensor]`, but `model_step` and helpers index it with string keys (`batch["params"]`, `batch["mel_spec"]`, `batch["noise"]`). The code works at runtime because the dataloader yields dicts, but the annotations are wrong.

**Impact:** High — pyright will flag every one of these once enabled. Misleading to readers.

**Fix:** Change all step signatures to accept the actual batch type:
```python
# Before
def training_step(self, batch: tuple[torch.Tensor, torch.Tensor], batch_idx: int) -> torch.Tensor:

# After
def training_step(self, batch: dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
```
Consider defining a `TypedDict` (e.g., `SurgeBatch`) if the batch shape is stable across modules.

---

## 2. `penalty` can be `None` → `TypeError` on `loss + penalty`

**Files affected:**
- `src/models/surge_flow_matching_module.py` — `_train_step` (L155)
- `src/models/ksin_flow_matching_module.py` — `training_step` (L371)

**Problem:** When `self.vector_field` has no `penalty()` method, `penalty` stays `None`. The code then returns `loss + penalty`, which raises `TypeError` at runtime.

**Impact:** Critical — this is a latent runtime crash, not just a typing issue. Any vector field without a `penalty()` method will crash training.

**Fix:**
```python
# Option A: default to zero tensor
penalty = getattr(self.vector_field, "penalty", lambda: torch.tensor(0.0, device=loss.device))()

# Option B: conditional return
return loss + penalty if penalty is not None else loss
```

---

## 3. Return type annotation syntax error

**File:** `scripts/rewrite_dataset_to_latest.py` (L10/L12)

**Problem:** `-> (Path, bool, str)` is a tuple *expression* (evaluates to a tuple of classes at runtime), not a type annotation. Type checkers ignore/misinterpret it.

**Impact:** Low — script code, but easy to fix.

**Fix:** `-> tuple[Path, bool, str]`

---

## 4. Suppressed `F821` (undefined name) hides real errors

**File:** `pyproject.toml` (L57-58, `per-file-ignores`)

**Problem:** `F821` is suppressed for `src/models/ksin_ff_module.py` and `src/models/surge_ff_module.py`, but both contain genuinely undefined references:
- `ksin_ff_module.py`: uses `MSESortLoss` (never imported/defined)
- `surge_ff_module.py`: references `SurgeFlowMatchingModule` in `__main__` block (never imported)

**Impact:** Medium — these are real bugs masked by the ignore rule. The `__main__` block one may be dead code, but `MSESortLoss` could indicate a missing import.

**Fix:** Fix or remove the undefined references, then drop the `F821` per-file-ignores.

---

## 5. Missing `hdf5plugin` import in `get_dataset_stats.py`

**File:** `scripts/get_dataset_stats.py` (L7)

**Problem:** The script reads HDF5 files via `h5py`, but doesn't import `hdf5plugin`. If shards were written with plugin compression (e.g., `Blosc2` as used in `src/data/vst/generate_vst_dataset.py`), reads will fail with "filter not available".

**Impact:** Medium — will crash on compressed datasets.

**Fix:** Add `import hdf5plugin  # noqa: F401` for side-effect registration.

---

## 6. Pre-commit config issues (non-typing, but flagged)

These are lower priority and partially addressed by the PR itself:

| Issue | File | Notes |
|-------|------|-------|
| `debug-statements` hook listed twice | `.pre-commit-config.yaml` | Remove duplicate |
| `validate-pyproject` configured twice | `.pre-commit-config.yaml` (L102) | Remove duplicate repo block |
| pyright `exclude` is very broad | `.pre-commit-config.yaml` | Intentional for now (incremental rollout per #49) |
| `.claude/settings.json` has local paths | `.claude/settings.json` | Already gitignored |

---

## Priority order

1. **penalty `None` bug** (#2) — latent runtime crash
2. **batch type annotations** (#1) — blocks pyright enablement
3. **F821 suppression** (#4) — masks real undefined names
4. **hdf5plugin import** (#5) — silent read failure on compressed data
5. **return type syntax** (#3) — cosmetic
6. **pre-commit duplicates** (#6) — cleanup

**Comment by mansibrahman03** [comment]
Hi! I'm a student in the CodePath AI301 program and I'd like to work on this issue. I'll start by reproducing the problem locally this week. Any pointers on where to start in the codebase would be appreciated!
